### PR TITLE
SEP-35: Operation IDs

### PR DESCRIPTION
**Pull request by debnil**
This protocol outlines an ID scheme for historical operations, so they can be consistently indexed across the ecosystem. It closes https://github.com/stellar/nebula/issues/18.

### CODE REVIEW

File: ecosystem/sep-0035.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+## Preamble
+
+```
+SEP: 0035
+Title: Operation IDs
+Author: Isaiah Turner (@Isaiah-Turner), Debnil Sur (@debnil)
```
Comment (bartekn) [inline ecosystem/sep-0035.md:6]: Can you add or mention Scott Fleckenstein - he actually invented this.

File: ecosystem/sep-0035.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+## Abstract
+This protocol defines an assignment method for IDs for operations in historical ledger data, representing the total order of operations on the Stellar network. The ledger sequence number, transaction application order, and operation index are packed into a single ID value.
+
+## Specification
+An ID is a single 64-bit signed integer. The sign guarantees SQL compatibility, since some dialects do not support unsigned integers.
+In order to create an id, three pieces of information are required. All three should be expressed in network byte order, which is big-endian. 
+1. **Operation Index**: the index of the operation within its parent transaction. This takes up the first 12 bits of the id. The maximum supported number of operations per transaction is 4095.
+2. **Transaction Application Order**: the order that the transaction was applied within the ledger where it was validated. This takes up the next 20 bits of the id. The maximum supported number of transactions per operation is 1,048,575.
+3. **Ledger Sequence**: this takes up the last 32 bits of the id. The maximum supported number of ledgers is 2,147,483,647.
```
Comment (bartekn) [inline ecosystem/sep-0035.md:27]: I think reverting the order (ledger, transaction, operation) will make it more readable.

File: ecosystem/sep-0035.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+[This diagram](https://github.com/stellar/go/blob/master/services/horizon/internal/toid/main.go#L22-L28) demonstrates the layout. The same link includes the reference implementation.
```
Comment (bartekn) [inline ecosystem/sep-0035.md:29]: Can we move the diagram to the SEP?

File: ecosystem/sep-0035.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+## Abstract
+This protocol defines an assignment method for IDs for operations in historical ledger data, representing the total order of operations on the Stellar network. The ledger sequence number, transaction application order, and operation index are packed into a single ID value.
+
+## Specification
+An ID is a single 64-bit signed integer. The sign guarantees SQL compatibility, since some dialects do not support unsigned integers.
+In order to create an id, three pieces of information are required. All three should be expressed in network byte order, which is big-endian. 
+1. **Operation Index**: the index of the operation within its parent transaction. This takes up the first 12 bits of the id. The maximum supported number of operations per transaction is 4095.
```
Comment (bartekn) [inline ecosystem/sep-0035.md:25]: Let's add the information that the index starts at 1.

File: ecosystem/sep-0035.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+2. **Transaction Application Order**: the order that the transaction was applied within the ledger where it was validated. This takes up the next 20 bits of the id. The maximum supported number of transactions per operation is 1,048,575.
```
Comment (bartekn) [inline ecosystem/sep-0035.md:26]: Let's add the information that the index starts at 1.

**Review by bartekn — APPROVED** [review]

File: ecosystem/sep-0035.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+0                   1                   2                   3
+0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                    Ledger Sequence Number                     |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|     Transaction Application Order     |       Op Index        |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+```
+
+1. **Ledger Sequence**: this takes up the last 32 bits of the id. The maximum supported number of ledgers is 2,147,483,647.
```
Comment (tomerweller) [inline ecosystem/sep-0035.md:36]: "last" 32 bits? aren't these first?

File: ecosystem/sep-0035.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+2. **Transaction Application Order**: the order that the transaction was applied within the ledger where it was validated. This takes up the middle 20 bits of the id. The index starts at 1. The maximum supported number of transactions per operation is 1,048,575.
```
Comment (tomerweller) [inline ecosystem/sep-0035.md:37]: "middle" seems somewhat opaque. maybe give the exact indices?

File: ecosystem/sep-0035.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+3. **Operation Index**: the index of the operation within its parent transaction. This takes up the first 12 bits of the id. The index starts at 1. The maximum supported number of operations per transaction is 4095.
```
Comment (tomerweller) [inline ecosystem/sep-0035.md:38]: following up on the first comment. Is this the first 12 bits or the last 12 bits?

**Comment by debnil** [comment]
@tomerweller: thanks for the double check. These comments were all spot on. I've specified the bit ranges, endianness, and left-padding to make it clearer.

**Pull request merged**